src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                        return TryConvertIfInMemberAccessExpression(source, (MemberAccessExpressionSyntax)parent, out documentUpdateInfo);
+                    // var a = new [] { from a in b select a };                    
+                    case SyntaxKind.ArrayInitializerExpression:
+                    // var a = from a in b select a
+                    case SyntaxKind.EqualsValueClause:
+                    // a = from a in b select a
+                    case SyntaxKind.SimpleAssignmentExpression:
+                    // (from a in b select a)?.ToList();
+                    case SyntaxKind.ConditionalAccessExpression:
+                    // new new List<IEnumerable<int>> { from a in q select a * a };
+                    case SyntaxKind.CollectionInitializerExpression:
+                    // new new List<int>(from a in q select a * a );
+                    case SyntaxKind.Argument:
+                        return TryReplaceWithLocalFunction(source, parent, out documentUpdateInfo);
+                    default:
+                        documentUpdateInfo = default;
+                        return false;
+                }
+            }
+
+            private bool TryConvertIfInMemberAccessExpression(
+               QueryExpressionSyntax source,
+               MemberAccessExpressionSyntax memberAccessExpression,
+               out DocumentUpdateInfo documentUpdateInfo)
+            {
+                if (memberAccessExpression.Parent is InvocationExpressionSyntax invocationExpression)
+                {
+                    // This also covers generic names (i.e. with type arguments) like 'ToList<int>'. 
+                    // The ValueText is still just 'ToList'. 
+                    switch (memberAccessExpression.Name.Identifier.ValueText)
+                    {
+                        case nameof(Enumerable.ToList):
+                            return TryConvertIfInToListInvocation(source, invocationExpression, out documentUpdateInfo);
+                        case nameof(Enumerable.Count):
+                            return TryConvertIfInCountInvocation(source, invocationExpression, out documentUpdateInfo);
+                        default:
+                            return TryReplaceWithLocalFunction(source, memberAccessExpression, out documentUpdateInfo);
+                    }
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool TryConvertIfInCountInvocation(
+                QueryExpressionSyntax source,
+                InvocationExpressionSyntax invocationExpression,
+                out DocumentUpdateInfo documentUpdateInfo)
+            {
+                if (_semanticModel.GetSymbolInfo(invocationExpression, _cancellationToken).Symbol is IMethodSymbol methodSymbol &&
+                    methodSymbol.Parameters.Length == 0 && methodSymbol.ReturnType?.SpecialType == SpecialType.System_Int32)
```

Comment:
>ReturnType [](start = 72, length = 10)

Should probably also check ref-ness of the return

Reply:
Do you mean !methodSymbol.ReturnsByRef?
Do you see any tests cases we should add?


---
In reply to: [176928958](https://github.com/dotnet/roslyn/pull/25362#discussion_r176928958) [](ancestors = 176928958)

Reply:
> Do you mean !methodSymbol.ReturnsByRef?

I would say ```RefKind == None```.

> Do you see any tests cases we should add?

Yes, each condition should be covered by negative an positive test

---
In reply to: [177586636](https://github.com/dotnet/roslyn/pull/25362#discussion_r177586636) [](ancestors = 177586636,176928958)

Reply:
I think we should check for ```RefKind == None``` to cover over kinds of ref-ness

---
In reply to: [177587626](https://github.com/dotnet/roslyn/pull/25362#discussion_r177587626) [](ancestors = 177587626,177586636,176928958)